File: Tests/Source/RangeTester.cpp
#include <catch2/catch.hpp>
#include "Aspen/Queue.hpp"
#include "Aspen/Range.hpp"
#include "Aspen/Shared.hpp"

using namespace Aspen;

TEST_CASE("test_backward_range", "[Range]") {
  auto reactor = range(constant(10), constant(9));
  REQUIRE(reactor.commit(0) == State::COMPLETE_EMPTY);
}

TEST_CASE("test_empty_range", "[Range]") {
  auto reactor = range(constant(1), constant(1));
  REQUIRE(reactor.commit(0) == State::COMPLETE_EMPTY);
}

TEST_CASE("test_single_range", "[Range]") {
  auto reactor = range(constant(1), constant(2));
  REQUIRE(reactor.commit(0) == State::COMPLETE_EVALUATED);
  REQUIRE(reactor.eval() == 1);
}

TEST_CASE("test_double_range", "[Range]") {
  auto reactor = range(constant(10), constant(12));
  REQUIRE(reactor.commit(0) == State::CONTINUE_EVALUATED);
  REQUIRE(reactor.eval() == 10);
  REQUIRE(reactor.commit(1) == State::COMPLETE_EVALUATED);
  REQUIRE(reactor.eval() == 11);
}

TEST_CASE("test_end_complete", "[Range]") {
  auto queue = Shared(Queue<int>());
  auto reactor = range(queue, constant(10));
  queue->push(8);
  REQUIRE(reactor.commit(0) == State::CONTINUE_EVALUATED);
  REQUIRE(reactor.eval() == 8);
  REQUIRE(reactor.commit(1) == State::COMPLETE_EVALUATED);
  REQUIRE(reactor.eval() == 9);
}

TEST_CASE("test_changing_start", "[Range]") {
  auto queue = Shared(Queue<int>());
  auto reactor = range(queue, constant(100));
  queue->push(8);
  REQUIRE(reactor.commit(0) == State::CONTINUE_EVALUATED);
  REQUIRE(reactor.eval() == 8);
  queue->push(50);
  REQUIRE(reactor.commit(1) == State::CONTINUE_EVALUATED);
  REQUIRE(reactor.eval() == 50);
  queue->push(20);
  REQUIRE(reactor.commit(2) == State::CONTINUE_EVALUATED);
  REQUIRE(reactor.eval() == 51);
  queue->push(200);
  REQUIRE(reactor.commit(3) == State::COMPLETE);
  REQUIRE(reactor.eval() == 51);
}

TEST_CASE("test_changing_end", "[Range]") {
  auto start_queue = Shared(Queue<int>());
  auto end_queue = Shared(Queue<int>());
  auto reactor = range(start_queue, end_queue);
  start_queue->push(0);
  end_queue->push(10);
  REQUIRE(reactor.commit(0) == State::CONTINUE_EVALUATED);
  REQUIRE(reactor.eval() == 0);
  REQUIRE(reactor.commit(1) == State::CONTINUE_EVALUATED);
  REQUIRE(reactor.eval() == 1);
  start_queue->push(6);
  end_queue->push(9);
  REQUIRE(reactor.commit(2) == State::CONTINUE_EVALUATED);
  REQUIRE(reactor.eval() == 6);
  end_queue->push(3);
  REQUIRE(reactor.commit(3) == State::NONE);
  end_queue->push(7);
  REQUIRE(reactor.commit(4) == State::NONE);
  end_queue->push(8);
  REQUIRE(reactor.commit(5) == State::EVALUATED);
  REQUIRE(reactor.eval() == 7);
  end_queue->set_complete();
  REQUIRE(reactor.commit(6) == State::COMPLETE);
}

File: Source/PythonTests/range_tester.py
import unittest

from fixtures import aspen, drive


class TestRange(unittest.TestCase):
  def test_ascending_range(self):
    reactor = aspen.range(0, 3)
    self.assertEqual(reactor.commit(0), aspen.State.CONTINUE_EVALUATED)
    self.assertEqual(reactor.eval(), 0)
    self.assertEqual(reactor.commit(1), aspen.State.CONTINUE_EVALUATED)
    self.assertEqual(reactor.eval(), 1)
    self.assertEqual(reactor.commit(2), aspen.State.COMPLETE_EVALUATED)
    self.assertEqual(reactor.eval(), 2)

  def test_descending_range(self):
    self.assertEqual(drive(aspen.range(3, 0, -1), 8), [3, 2, 1])

  def test_step(self):
    self.assertEqual(drive(aspen.range(0, 10, 3), 8), [0, 3, 6, 9])

  def test_stop_below_the_start(self):
    self.assertEqual(aspen.range(10, 9).commit(0), aspen.State.COMPLETE)

  def test_fractional_step(self):
    self.assertEqual(drive(aspen.range(0.0, 1.0, 0.5), 8), [0.0, 0.5])


if __name__ == '__main__':
  unittest.main()
